### ISSUE TITLE
[YSQL] INSERT ON CONFLICT is slow when there are multiple arbiter indexes

### ISSUE THREAD

**Issue by karthik-ramanathan-3006** [yugabyte/yugabyte-db#23115]
### Description

INSERT ON CONFLICT performs constraint checking (index or exclusion) in sequence.
If there are multiple candidate constraints to be checked, then YSQL initializes and performs an index scan for indexes associated with each constraint.

Consider the following setup:
```
CREATE TABLE ioc (a INT PRIMARY KEY, b INT, c INT, d INT);
CREATE UNIQUE INDEX NONCONCURRENTLY ON ioc (b);
CREATE UNIQUE INDEX NONCONCURRENTLY ON ioc (b);

-- Insert a single row into the table
INSERT INTO ioc VALUES (1, 1, 1, 1);

-- For brevity of output
SET yb_explain_hide_non_deterministic_fields TO true;
```

**Table description**
yugabyte=# \d ioc
                Table "public.ioc"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 a      | integer |           | not null |
 b      | integer |           |          |
 c      | integer |           |          |
 d      | integer |           |          |
Indexes:
    "ioc_pkey" PRIMARY KEY, lsm (a HASH)
    "ioc_b_idx" UNIQUE, lsm (b HASH)
    "ioc_b_idx1" UNIQUE, lsm (b HASH)
```

**Case 1: ON CONFLICT DO NOTHING -- no arbiter columns specified**
```
yugabyte=# EXPLAIN (ANALYZE, DIST) INSERT INTO ioc VALUES (2, 2, 2, 2) ON CONFLICT DO NOTHING;
                                QUERY PLAN
--------------------------------------------------------------------------
 Insert on ioc  (cost=0.00..0.01 rows=1 width=16) (actual rows=0 loops=1)
   Conflict Resolution: NOTHING
   Tuples Inserted: 1
   Conflicting Tuples: 0
   ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual rows=1 loops=1)
         Storage Table Read Requests: 1
         Storage Index Read Requests: 2
         Storage Table Write Requests: 1
         Storage Index Write Requests: 2
 Storage Read Requests: 3
 Storage Rows Scanned: 0
 Storage Write Requests: 3
 Storage Flush Requests: 1
(13 rows)

yugabyte=# DELETE FROM ioc WHERE a = 2;
DELETE 1
```
To check if the row (2, 2, 2, 2) violates any constraints, each of the indexes `ioc_pkey`, `ioc_b_idx`, `ioc_b_idx1` are read in sequence. 
Three issues here:
1. The indexes are read in sequence, which can prove to be expensive
2. `Conflict Arbiter Indexes` are not displayed
3. It should be possible to infer that scanning both `ioc_b_idx`, `ioc_b_idx1` is redundant. This is because these are unique non-partial indexes. The same may not apply if they are exclusion indexes or partial indexes.


**Case 2: ON CONFLICT DO NOTHING -- arbiter column(s) specified**
```
yugabyte=# EXPLAIN (ANALYZE, DIST) INSERT INTO ioc VALUES (2, 2, 2, 2) ON CONFLICT (b) DO NOTHING;
                                QUERY PLAN
--------------------------------------------------------------------------
 Insert on ioc  (cost=0.00..0.01 rows=1 width=16) (actual rows=0 loops=1)
   Conflict Resolution: NOTHING
   Conflict Arbiter Indexes: ioc_b_idx, ioc_b_idx1
   Tuples Inserted: 1
   Conflicting Tuples: 0
   ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual rows=1 loops=1)
         Storage Index Read Requests: 2
         Storage Table Write Requests: 1
         Storage Index Write Requests: 2
 Storage Read Requests: 2
 Storage Rows Scanned: 0
 Storage Write Requests: 3
 Storage Flush Requests: 1
(13 rows)

yugabyte=# DELETE FROM ioc WHERE a = 2;
DELETE 1
```

The observations in Case 1 apply here as well.

**Case 3: ON CONFLICT DO UPDATE -- arbiter column(s) specified**
Note that ON CONFLICT DO UPDATE does not allow skipping of arbiter columns (ie. "automatic" inference of arbiter indexes is not allowed)

```
yugabyte=# EXPLAIN (ANALYZE, DIST) INSERT INTO ioc VALUES (2, 2, 2, 2) ON CONFLICT (b) DO UPDATE SET d = EXCLUDED.d;
                                QUERY PLAN
--------------------------------------------------------------------------
 Insert on ioc  (cost=0.00..0.01 rows=1 width=16) (actual rows=0 loops=1)
   Conflict Resolution: UPDATE
   Conflict Arbiter Indexes: ioc_b_idx, ioc_b_idx1
   Tuples Inserted: 0
   Conflicting Tuples: 1
   ->  Result  (cost=0.00..0.01 rows=1 width=16) (actual rows=1 loops=1)
         Storage Table Read Requests: 1
         Storage Table Rows Scanned: 1
         Storage Index Read Requests: 1
         Storage Index Rows Scanned: 1
         Storage Table Write Requests: 1
         Storage Index Write Requests: 4
         Storage Flush Requests: 1
 Storage Read Requests: 2
 Storage Rows Scanned: 2
 Storage Write Requests: 5
 Storage Flush Requests: 2
(17 rows)
```
Since only UNIQUE-ness constraints are supported, it is guaranteed that if there is a conflict, it will be found in the first index that is scanned. A couple of issues can potentially be surfaced when EXCLUSION constraints are supported:
1. The indexes will be read in sequence, and this can be expensive.
2. The slot holding the conflicting tuple may be lost, resulting in an additional scan to be fetched again. (To the best of my understanding, this is already handled, but it would be good to check this again)

In summary, this issue highlights the following:
- Scanning indexes for uniqueness or exclusion constraints in sequence can be a bottleneck
- When only unique indexes are involved, it should be possible to eliminate redundant scans 



### Issue Type

kind/enhancement

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.
